### ISSUE TITLE
LBFGS optimizer

### ISSUE THREAD

**Issue by qzhu2017** [denizyuret/Knet.jl#590]
Although I know most of people use ADAM or SGD for the optimization of weight in NN. However, for some small NN architecture, the optimization method with line search (e.g., LBFGS) would be far more efficient. I wonder if the Knet developers would be interested in implementing LBFGS?

**Comment by denizyuret** [comment]
Optim.jl may already have this.

**Comment by qzhu2017** [comment]
Yes, I know. Can I using it with Kent? I tried it. But this is not very obvious.

**Comment by denizyuret** [comment]
```
x = Param(Array(...))
J = @diff f(x)
g = grad(J,x)
```
Should give you a `g` with the exact type/shape as `x`. See `@doc AutoGrad`.
